File: cpp/src/adaptive_metropolis.cpp
/**
 * @file adaptive_metropolis.cpp
 * @brief Implementation of Adaptive Metropolis-Hastings MCMC sampler
 */

#include "adaptive_metropolis.h"
#include <iostream>
#include <iomanip>
#include <chrono>
#include <algorithm>
#include <cmath>

AdaptiveMetropolis::AdaptiveMetropolis(
    LogProbFunction log_prob_fn,
    int n_params,
    const std::vector<std::string>& param_names,
    double initial_step_size,
    int adaptation_interval,
    BoundsFunction bounds_check
) : log_prob_fn_(log_prob_fn),
    bounds_check_(bounds_check),
    n_params_(n_params),
    param_names_(param_names),
    step_size_(initial_step_size),
    adaptation_interval_(adaptation_interval),
    epsilon_(1e-3),  // Standard epsilon for Haario et al.
    adaptation_start_(std::max(100, 2 * n_params)),
    target_accept_rate_(0.234),
    normal_dist_(0.0, 1.0),
    uniform_dist_(0.0, 1.0) {
    
    // Initialize parameter names if not provided
    if (param_names_.empty()) {
        param_names_.resize(n_params_);
        for (int i = 0; i < n_params_; ++i) {
            param_names_[i] = "param_" + std::to_string(i);
        }
    }
    
    // Initialize covariance matrix as scaled identity (Haario et al.)
    // Use the provided step size as the initial scale
    double scale = initial_step_size * initial_step_size;
    cov_matrix_ = Eigen::MatrixXd::Identity(n_params_, n_params_) * scale;
}

MCMCResult AdaptiveMetropolis::sample(
    const Eigen::VectorXd& initial_params,
    int n_samples,
    int n_warmup,
    int thin,
    unsigned int seed
) {
    auto start_time = std::chrono::high_resolution_clock::now();
    
    rng_.seed(seed);
    
    int total_iterations = n_warmup + n_samples * thin;
    
    // Storage for all samples (including warmup for adaptation)
    std::vector<Eigen::VectorXd> all_samples;
    std::vector<Eigen::VectorXd> normal_proposals_only;  // Only non-flip samples for covariance
    std::vector<double> all_log_probs;
    all_samples.reserve(total_iterations);
    normal_proposals_only.reserve(total_iterations);
    all_log_probs.reserve(total_iterations);
    
    // Current state
    Eigen::VectorXd current_params = initial_params;
    double current_log_prob = log_prob_fn_(current_params);
    
    // Check initial state
    if (!std::isfinite(current_log_prob)) {
        throw std::runtime_error("Initial log probability is not finite: " + std::to_string(current_log_prob));
    }
    
    if (!check_bounds(current_params)) {
        throw std::runtime_error("Initial parameters are outside bounds");
    }
    
    // Counters
    int n_accepted = 0;
    
    std::cout << "🎯 Starting Adaptive Metropolis: " << n_samples << " samples + " << n_warmup << " warmup" << std::endl;
    std::cout << "Initial log probability: " << std::fixed << std::setprecision(6) << current_log_prob << std::endl;
    
    // Progress tracking
    int progress_interval = std::max(1, total_iterations / 100);
    
    // Main sampling loop
    for (int i = 0; i < total_iterations; ++i) {
        // Progress indicator
        if (i % progress_interval == 0) {
            int progress = (i * 100) / total_iterations;
            std::cout << "\rMCMC sampling: " << progress << "%" << std::flush;
        }
        
        // Generate proposal (track if it's a flip for covariance calculation)
        bool is_flip_proposal;
        Eigen::VectorXd proposed_params = propose_step(current_params, is_flip_proposal);
        
        // Check bounds - if out of bounds, reject immediately
        if (!check_bounds(proposed_params)) {
            // Reject proposal - stay at current state
            // This is the standard MCMC approach for handling bounds
        } else {
            try {
                // Evaluate log probability at proposal
                double proposed_log_prob = log_prob_fn_(proposed_params);
                
                if (std::isfinite(proposed_log_prob)) {
                    // Metropolis acceptance criterion
                    double log_alpha = proposed_log_prob - current_log_prob;
                    double alpha = std::exp(std::min(0.0, log_alpha));
                    
                    // Accept or reject
                    if (uniform_dist_(rng_) < alpha) {
                        current_params = proposed_params;
                        current_log_prob = proposed_log_prob;
                        n_accepted++;
                        
                        // Only store non-flip proposals for covariance adaptation
                        if (!is_flip_proposal) {
                            normal_proposals_only.push_back(current_params);
                        }
                    } else {
                        // Rejected - still store current state if it came from normal proposal
                        if (!is_flip_proposal) {
                            normal_proposals_only.push_back(current_params);
                        }
                    }
                }
            } catch (const std::exception& e) {
                // If evaluation fails, reject proposal
            }
        }
        
        // Store sample
        all_samples.push_back(current_params);
        all_log_probs.push_back(current_log_prob);
        
        // Adaptation during warmup - use only normal proposals for covariance
        if (i < n_warmup && (i + 1) % adaptation_interval_ == 0 && normal_proposals_only.size() > 50) {
            // Build matrix from normal proposals only
            int n_normal = normal_proposals_only.size();
            int start_idx = std::max(0, n_normal - 1000);  // Use last 1000 normal proposals
            
            Eigen::MatrixXd normal_samples(n_normal - start_idx, n_params_);
            for (int j = start_idx; j < n_normal; ++j) {
                normal_samples.row(j - start_idx) = normal_proposals_only[j];
            }
            
            adapt_covariance(normal_samples, i + 1);
            
            // Adapt step size
            double recent_accept_rate = static_cast<double>(n_accepted) / (i + 1);
            adapt_step_size(recent_accept_rate, i + 1);
        }
    }
    
    std::cout << std::endl; // New line after progress
    
    // Extract final samples (post-warmup, thinned)
    Eigen::MatrixXd samples(n_samples, n_params_);
    Eigen::VectorXd log_probs(n_samples);
    
    int sample_idx = 0;
    for (int i = n_warmup; i < total_iterations && sample_idx < n_samples; i += thin) {
        samples.row(sample_idx) = all_samples[i];
        log_probs(sample_idx) = all_log_probs[i];
        sample_idx++;
    }
    
    // Calculate final acceptance rate
    double final_accept_rate = static_cast<double>(n_accepted) / total_iterations;
    
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
    double total_time = duration.count() / 1000.0;
    
    std::cout << "✅ Sampling completed!" << std::endl;
    std::cout << "   Acceptance rate: " << std::fixed << std::setprecision(3) << final_accept_rate << std::endl;
    std::cout << "   Final step size: " << std::fixed << std::setprecision(6) << step_size_ << std::endl;
    std::cout << "   Total time: " << std::fixed << std::setprecision(2) << total_time << " seconds" << std::endl;
    
    return MCMCResult{
        samples,
        log_probs,
        final_accept_rate,
        param_names_,
        cov_matrix_,
        step_size_,
        total_time
    };
}

Eigen::VectorXd AdaptiveMetropolis::propose_step(const Eigen::VectorXd& current_params, bool& is_flip) {
    // Decide whether to do mass ordering flip (50% chance - flat prior on mass ordering)
    std::uniform_real_distribution<double> uniform(0.0, 1.0);
    is_flip = uniform(rng_) < 0.5;
    
    if (is_flip && current_params.size() >= 6) {
        // Mass ordering flip: just flip dm32 sign (parameter 5)
        Eigen::VectorXd flipped_params = current_params;
        flipped_params(5) = -current_params(5);  // Flip dm32
        return flipped_params;
    } else {
        // Not actually a flip if we don't have enough parameters
        is_flip = false;
        // Standard Haario et al. multivariate normal proposal
        Eigen::VectorXd random_step(n_params_);
        for (int i = 0; i < n_params_; ++i) {
            random_step(i) = normal_dist_(rng_);
        }
        
        // Use Cholesky decomposition for multivariate normal
        Eigen::LLT<Eigen::MatrixXd> llt(cov_matrix_);
        if (llt.info() == Eigen::Success) {
            random_step = llt.matrixL() * random_step;
        } else {
            // Fallback: use sqrt of diagonal elements
            for (int i = 0; i < n_params_; ++i) {
                random_step(i) *= std::sqrt(std::max(epsilon_, cov_matrix_(i, i)));
            }
        }
        
        Eigen::VectorXd proposed_params = current_params + random_step;
        
        // Apply periodic boundary condition to deltacp (parameter 3) if it exists
        if (proposed_params.size() >= 4) {
            const double two_pi = 2.0 * M_PI;
            while (proposed_params(3) < 0.0) proposed_params(3) += two_pi;
            while (proposed_params(3) >= two_pi) proposed_params(3) -= two_pi;
        }
        
        return proposed_params;
    }
}

void AdaptiveMetropolis::adapt_covariance(const Eigen::MatrixXd& samples, int iteration) {
    if (iteration < adaptation_start_) {
        return;
    }
    
    // Compute empirical covariance
    Eigen::VectorXd sample_mean = samples.colwise().mean();
    Eigen::MatrixXd centered_samples = samples.rowwise() - sample_mean.transpose();
    Eigen::MatrixXd empirical_cov = centered_samples.transpose() * centered_samples / (samples.rows() - 1);
    
    // Adaptive scaling factor (Haario et al. 2001)
    double s_d = 2.38 * 2.38 / n_params_;  // This is the correct scaling
    
    // More aggressive adaptation early on, then stabilize
    double adaptation_weight = 1.0;
    if (iteration < 1000) {
        adaptation_weight = 0.8;  // Faster adaptation early
    } else if (iteration < 5000) {
        adaptation_weight = 0.5;  // Medium adaptation
    } else {
        adaptation_weight = 0.1;  // Conservative adaptation later
    }
    
    // Weighted update of covariance matrix
    Eigen::MatrixXd target_cov = s_d * (empirical_cov + epsilon_ * Eigen::MatrixXd::Identity(n_params_, n_params_));
    cov_matrix_ = (1.0 - adaptation_weight) * cov_matrix_ + adaptation_weight * target_cov;
}

void AdaptiveMetropolis::adapt_step_size(double accept_rate, int iteration) {
    if (iteration < 100) { // Don't adapt too early
        return;
    }
    
    // More aggressive adaptation rate early on, then reduce
    double adaptation_rate;
    if (iteration < 1000) {
        adaptation_rate = 0.2;  // More aggressive early adaptation
    } else if (iteration < 5000) {
        adaptation_rate = 0.1;
    } else {
        adaptation_rate = std::min(0.05, 100.0 / iteration);  // Gentler later
    }
    
    // Dual averaging for step size (more stable than multiplicative)
    double error = accept_rate - target_accept_rate_;
    step_size_ = step_size_ * std::exp(adaptation_rate * error);
    
    // Keep step size reasonable
    step_size_ = std::max(1e-6, std::min(step_size_, 0.5));
    
    // Don't rescale covariance matrix here - let it adapt independently
}

bool AdaptiveMetropolis::check_bounds(const Eigen::VectorXd& params) {
    if (bounds_check_) {
        return bounds_check_(params);
    }
    return true; // No bounds checking if not provided
}

File: src/magpy/mcmc/adaptive_metropolis.py
"""
Adaptive Metropolis MCMC sampler for neutrino oscillation parameter inference.

This module implements an adaptive Metropolis-Hastings algorithm with:
- Adaptive proposal covariance matrix
- Parameter bounds checking
- Automatic step size adaptation
- Diagnostic statistics
"""

import jax.numpy as jnp
from jax import random
from typing import Callable, List, Tuple, Optional
from dataclasses import dataclass
from tqdm import tqdm


@dataclass
class MCMCResult:
    """Results from MCMC sampling."""
    samples: jnp.ndarray
    log_prob: jnp.ndarray
    accept_rate: float
    param_names: List[str]
    covariance_matrix: jnp.ndarray
    final_step_size: float


class AdaptiveMetropolis:
    """
    Adaptive Metropolis-Hastings sampler.
    
    Implements the adaptive Metropolis algorithm from Haario et al. (2001)
    with parameter bounds checking and automatic step size adaptation.
    """
    
    def __init__(
        self,
        log_prob_fn,
        n_params: int,
        param_names=None,
        initial_step_size: float = 0.1,
        adaptation_interval: int = 100,
        bounds=None
    ):
        """
        Initialize the adaptive Metropolis sampler.
        
        Args:
            log_prob_fn: Function that computes log probability
            n_params: Number of parameters
            param_names: Parameter names for diagnostics
            initial_step_size: Initial step size for proposals
            adaptation_interval: How often to adapt covariance matrix
            bounds: Parameter bounds as list of (min, max) tuples
        """
        self.log_prob_fn = log_prob_fn
        self.n_params = n_params
        self.param_names = param_names or [f"param_{i}" for i in range(n_params)]
        self.step_size = initial_step_size
        self.adaptation_interval = adaptation_interval
        self.bounds = bounds
        
        # Adaptive parameters
        self.epsilon = 1e-8  # Small regularization
        self.adaptation_start = max(100, 2 * n_params)  # When to start adapting
        
        # Initialize covariance matrix as identity
        self.cov_matrix = jnp.eye(n_params) * (initial_step_size ** 2)
        
        # Target acceptance rate
        self.target_accept_rate = 0.234  # Optimal for high-dimensional problems
        
    def check_bounds(self, params: jnp.ndarray) -> bool:
        """Check if parameters are within bounds."""
        if self.bounds is None:
            return True
            
        for i, (low, high) in enumerate(self.bounds):
            if params[i] < low or params[i] > high:
                return False
        return True
        
    def propose_step(self, current_params: jnp.ndarray, key: jnp.ndarray) -> jnp.ndarray:
        """Generate proposal step using current covariance matrix."""
        # Generate random step from multivariate normal
        proposal_step = random.multivariate_normal(
            key, 
            mean=jnp.zeros(self.n_params),
            cov=self.cov_matrix
        )
        
        return current_params + proposal_step
        
    def adapt_covariance(self, samples: jnp.ndarray, iteration: int):
        """Adapt the covariance matrix based on sample history."""
        if iteration < self.adaptation_start:
            return
            
        # Compute empirical covariance
        sample_mean = jnp.mean(samples, axis=0)
        centered_samples = samples - sample_mean
        empirical_cov = jnp.cov(centered_samples.T)
        
        # Adaptive scaling factor (Haario et al. 2001)
        s_d = 2.38**2 / self.n_params
        
        # Update covariance matrix with regularization
        self.cov_matrix = s_d * (empirical_cov + self.epsilon * jnp.eye(self.n_params))
        
    def adapt_step_size(self, accept_rate: float, iteration: int):
        """Adapt step size based on acceptance rate."""
        if iteration < 100:  # Don't adapt too early
            return
            
        # Simple step size adaptation
        adaptation_rate = min(0.1, 1.0 / iteration**0.6)
        
        if accept_rate > self.target_accept_rate:
            self.step_size *= (1 + adaptation_rate)
        else:
            self.step_size *= (1 - adaptation_rate)
            
        # Keep step size reasonable
        self.step_size = jnp.clip(self.step_size, 1e-8, 1.0)
        
        # Update covariance matrix scaling
        scale_factor = (self.step_size / 0.1) ** 2
        self.cov_matrix = self.cov_matrix * scale_factor
        
    def sample(
        self,
        initial_params,
        n_samples: int,
        n_warmup: int = 1000,
        key=None,
        thin: int = 1
    ):
        """
        Run adaptive Metropolis sampling.
        
        Args:
            initial_params: Starting parameter values
            n_samples: Number of samples to collect (after thinning)
            n_warmup: Number of warmup samples for adaptation
            key: Random key
            thin: Thinning factor (keep every thin-th sample)
            
        Returns:
            MCMCResult with samples and diagnostics
        """
        if key is None:
            key = random.PRNGKey(42)
        
        assert key is not None  # Type hint for mypy
            
        # Total iterations needed
        total_iterations = n_warmup + n_samples * thin
        
        # Storage for all samples (including warmup for adaptation)
        all_samples = []
        all_log_probs = []
        
        # Current state
        current_params = initial_params.copy()
        current_log_prob = self.log_prob_fn(current_params)
        
        # Check if initial parameters are valid
        if not jnp.isfinite(current_log_prob):
            raise ValueError(f"Initial log probability is not finite: {current_log_prob}")
            
        if not self.check_bounds(current_params):
            raise ValueError("Initial parameters are outside bounds")
        
        # Counters
        n_accepted = 0
        
        print(f"🎯 Starting Adaptive Metropolis: {n_samples} samples + {n_warmup} warmup")
        print(f"Initial log probability: {current_log_prob:.6f}")
        
        # Main sampling loop
        for i in tqdm(range(total_iterations), desc="MCMC sampling"):
            key, subkey = random.split(key)
            
            # Generate proposal
            proposed_params = self.propose_step(current_params, subkey)
            
            # Check bounds
            if not self.check_bounds(proposed_params):
                # Reject proposal - stays at current state
                pass
            else:
                # Evaluate log probability at proposal
                try:
                    proposed_log_prob = self.log_prob_fn(proposed_params)
                    
                    if jnp.isfinite(proposed_log_prob):
                        # Metropolis acceptance criterion
                        log_alpha = proposed_log_prob - current_log_prob
                        alpha = jnp.exp(jnp.minimum(0.0, log_alpha))
                        
                        # Accept or reject
                        key, subkey = random.split(key)
                        if random.uniform(subkey) < alpha:
                            current_params = proposed_params
                            current_log_prob = proposed_log_prob
                            n_accepted += 1
                except Exception:
                    # If evaluation fails, reject proposal
                    pass
            
            # Store sample
            all_samples.append(current_params.copy())
            all_log_probs.append(current_log_prob)
            
            # Adaptation during warmup
            if i < n_warmup and (i + 1) % self.adaptation_interval == 0:
                # Adapt covariance matrix
                recent_samples = jnp.array(all_samples[max(0, i - 500):i + 1])
                self.adapt_covariance(recent_samples, i + 1)
                
                # Adapt step size
                recent_accept_rate = n_accepted / (i + 1)
                self.adapt_step_size(recent_accept_rate, i + 1)
        
        # Extract final samples (post-warmup, thinned)
        warmup_samples = all_samples[n_warmup:]
        warmup_log_probs = all_log_probs[n_warmup:]
        
        # Apply thinning
        final_samples = warmup_samples[::thin][:n_samples]
        final_log_probs = warmup_log_probs[::thin][:n_samples]
        
        # Convert to arrays
        samples = jnp.array(final_samples)
        log_probs = jnp.array(final_log_probs)
        
        # Calculate final acceptance rate
        final_accept_rate = n_accepted / total_iterations
        
        print(f"✅ Sampling completed!")
        print(f"   Acceptance rate: {final_accept_rate:.3f}")
        print(f"   Final step size: {self.step_size:.6f}")
        
        return MCMCResult(
            samples=samples,
            log_prob=log_probs,
            accept_rate=final_accept_rate,
            param_names=self.param_names,
            covariance_matrix=self.cov_matrix,
            final_step_size=float(self.step_size)
        )
